api/client/proxy/transport/transportv1/client.go

```diff
@@ -0,0 +1,228 @@
+		return nil, trail.FromGRPC(err, "failed to send cluster request")
+	}
+
+	streamRW, err := streamutils.NewReadWriter(clusterStream{stream: stream})
+	if err != nil {
+		return nil, trace.Wrap(err, "unable to create stream reader")
+	}
+
+	p, ok := peer.FromContext(stream.Context())
+	if !ok {
+		return nil, trace.BadParameter("unable to retrieve peer information")
+	}
+
+	return streamutils.NewConn(streamRW, src, p.Addr), nil
+}
+
+// clusterStream implements the [streamutils.Source] interface
+// for a [proxyv1.ProxyService_ProxyClusterClient].
+type clusterStream struct {
+	stream proxyv1.ProxyService_ProxyClusterClient
+}
+
+func (c clusterStream) Recv() ([]byte, error) {
+	req, err := c.stream.Recv()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if req.Frame == nil {
+		return nil, trace.BadParameter("received invalid frame")
+	}
+
+	return req.Frame.Payload, nil
+}
+
+func (c clusterStream) Send(frame []byte) error {
+	return trace.Wrap(c.stream.Send(&proxyv1.ProxyClusterRequest{Frame: &proxyv1.Frame{Payload: frame}}))
+}
+
+// DialHost establishes a connection to the instance in the provided cluster that matches
+// the hostport. If a keyring is provided then it will be forwarded to the remote instance.
+// The src address will be used as the LocalAddr of the returned [net.Conn].
+func (c *Client) DialHost(ctx context.Context, hostport, cluster string, src net.Addr, keyring agent.ExtendedAgent) (net.Conn, *proxyv1.ClusterDetails, error) {
+	stream, err := c.clt.ProxySSH(ctx)
+	if err != nil {
+		return nil, nil, trail.FromGRPC(err, "unable to establish proxy stream")
+	}
+
+	if err := stream.Send(&proxyv1.ProxySSHRequest{DialTarget: &proxyv1.TargetHost{
+		HostPort: hostport,
+		Cluster:  cluster,
+	}}); err != nil {
+		return nil, nil, trail.FromGRPC(err, "failed to send dial target request")
+	}
+
+	resp, err := stream.Recv()
+	if err != nil {
+		return nil, nil, trail.FromGRPC(err, "failed to receive cluster details response")
+	}
+
+	// create a stream for agent protocol
+	agentStream := newClientStream(stream, func(payload []byte) *proxyv1.ProxySSHRequest {
+		return &proxyv1.ProxySSHRequest{Frame: &proxyv1.ProxySSHRequest_Agent{Agent: &proxyv1.Frame{Payload: payload}}}
+	})
+	// create a reader writer for agent protocol
+	agentRW, err := streamutils.NewReadWriter(agentStream)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	// create a stream for ssh protocol
+	sshStream := newClientStream(stream, func(payload []byte) *proxyv1.ProxySSHRequest {
+		return &proxyv1.ProxySSHRequest{Frame: &proxyv1.ProxySSHRequest_Ssh{Ssh: &proxyv1.Frame{Payload: payload}}}
+	})
+
+	// create a reader writer for SSH protocol
+	sshRW, err := streamutils.NewReadWriter(sshStream)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	// multiplex the frames to the correct handlers
+	var serveOnce sync.Once
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
```

Comment:
This err check treats an `io.EOF` the same as any other error - is there a missing behaviour here for an error that isn't an `io.EOF` ?

Reply:
Do we want to consider warn/debug logging for these ?